### DISCUSSION TITLE
adding custom argument to streamlink

### DISCUSSION THREAD

**Discussion by Rob2k9** [streamlink/streamlink#5071] [Q&A]
I am using streamlink to playback content from my nginx rtmp server over the local network i have 33 pcs connected to the network

my nginx rtmp server outputs a dash stream but authentication is handeled by a 3rd party code provided to each person i have a gui application in python that calls streamlink to play one of the dash streams from my nginx server i would like each  person to pass there own code to the application and the application then passes the code back to streamlink to check before playback 

i have edited - argparser.py and added a --sec-key option to this file i have also edited session.py and added the same --sec-key option in here i then went to ffmpegmux.py and edited this file 

i added in the following

seckey = session.options.get("sec-key") or options.pop("seckey", None)

and then checked to see if seckey was set and passed it off to a simple get request to http://0.0.0.0/sec.php?key=4564654

this would return true or false via json if the key was active or not

i thought that was everyi could see todo with passing a custom argument from cli but when i run streamlink i get the error "unrecognised argument: --sec-key 4564654"

the reason iam using streamlink and piping the output is when playback is done in vlc or a browser player the dash playback seems to be choppy at best but using streamlink it plays perfect it also allows me to provide simple and clean interface for playback with no complicated options

have i missed something when trying to pass a custom argument to streamlink ?

**Comment by mkbloke** [comment]
Rather than modifying Streamlink, I'm thinking it could be easier to do it another way using existing functionality.

If you were to pass the key as an HTTP header, then you could just start Streamlink via:
```
streamlink --http-header=X-Security-Key=123456 <URL> <STREAM>
```
If the code changes you'd just need a simple wrapper script/GUI to take the code input from the user and pass it to Streamlink's command line when you run it.

**Reply by mkbloke** [reply]
Another thought on this: if you must embed the code as part of the URL, why not do it via a variable substitution on the request URL?

**Reply by Rob2k9** [reply]
well i wanted to edit streamlink because the password the users gets is modified in python before being attached to the url this is done in the android app that is provided for people on the local network as well but unlike the android app its a lot harder to obsf python code to stop people reading and seeing what is happening to add it to streamlink was an extra step to stop any users actually seeing what was happening

adding it asa header would require modification of the nginx auth code as well as modification of the android app streamlink mofification just seemed like the most obvs and easy option

right now i have tested with a static password being used and streamlink works fine and the stream plays in the player no problem at all litrully the only problem i have is i cant get streamlink to accept the custom argument even though it seems like i do have everything in the correct place for it to work 

it just seems like when i run streamlink its using a cached version of the code or something and not actually picking up the changes to args or i missed something

i have looked over the code of streamlink and the cli options even the modifications to edits over the life of the project to see how other args have been added and i tried to recreate this as best as i possibly could

**Comment by mkbloke** [comment]
This works for me:
```
diff --git a/src/streamlink/stream/ffmpegmux.py b/src/streamlink/stream/ffmpegmux.py
index 7a0dfacb..bfdf5d89 100644
--- a/src/streamlink/stream/ffmpegmux.py
+++ b/src/streamlink/stream/ffmpegmux.py
@@ -153,6 +153,7 @@ class FFMPEGMuxer(StreamIO):
             raise StreamError("cannot use FFMPEG")
 
         self.session = session
+        log.info(f'KEY={session.options.get("sec-key")}')
         self.process = None
         self.streams = streams
 
diff --git a/src/streamlink_cli/argparser.py b/src/streamlink_cli/argparser.py
index 4fac5501..843060de 100644
--- a/src/streamlink_cli/argparser.py
+++ b/src/streamlink_cli/argparser.py
@@ -375,6 +375,14 @@ def build_parser():
         Resolve address names to IPv6 only. This option overrides --ipv4.
         """
     )
+    general.add_argument(
+        "--sec-key",
+        type=int,
+        metavar="KEY",
+        help="""
+        Provide a security key.
+        """
+    )
 
     player = parser.add_argument_group("Player options")
     player.add_argument(
@@ -1226,6 +1234,7 @@ def false(_): return False  # noqa: E704
 _ARGUMENT_TO_SESSIONOPTION: List[Tuple[str, str, Optional[Callable[[Any], Any]]]] = [
     # generic arguments
     ("locale", "locale", None),
+    ("sec_key", "sec-key", None),
 
     # network arguments
     ("interface", "interface", None),
```
```
$ streamlink 'https://bitmovin-a.akamaihd.net/content/MI201109210084_1/mpds/f08e80da-bf1d-4e3d-8899-f0f6155f6efa.mpd' 270p
[cli][info] Found matching plugin dash for URL https://bitmovin-a.akamaihd.net/content/MI201109210084_1/mpds/f08e80da-bf1d-4e3d-8899-f0f6155f6efa.mpd
[cli][info] Available streams: 180p (worst), 270p, 360p, 540p, 720p, 1080p (best)
[cli][info] Opening stream: 270p (dash)
[cli][info] Starting player: mpv
[stream.ffmpegmux][info] KEY=None
[utils.named_pipe][info] Creating pipe streamlinkpipe-918578-1-2317
[utils.named_pipe][info] Creating pipe streamlinkpipe-918578-2-7151
[cli][info] Stream ended
[cli][info] Closing currently open stream...

$ streamlink --sec-key 123456 'https://bitmovin-a.akamaihd.net/content/MI201109210084_1/mpds/f08e80da-bf1d-4e3d-8899-f0f6155f6efa.mpd' 270p
[cli][info] Found matching plugin dash for URL https://bitmovin-a.akamaihd.net/content/MI201109210084_1/mpds/f08e80da-bf1d-4e3d-8899-f0f6155f6efa.mpd
[cli][info] Available streams: 180p (worst), 270p, 360p, 540p, 720p, 1080p (best)
[cli][info] Opening stream: 270p (dash)
[cli][info] Starting player: mpv
[stream.ffmpegmux][info] KEY=123456
[utils.named_pipe][info] Creating pipe streamlinkpipe-918669-1-1777
[utils.named_pipe][info] Creating pipe streamlinkpipe-918669-2-6329
[cli][info] Stream ended
[cli][info] Closing currently open stream...
```
